docs/DdnRelocatableShaderElf.md

```diff
@@ -0,0 +1,227 @@
+similar to loading descriptor set 1 at binding 0. Then the offset of the
+descriptor is added to that.
+
+An unlinked shader compilation has to account for both a top level descriptor or
+a descriptor in a descriptor table. An example can be found in the
+[buffer descriptor load test](../llpc/test/shaderdb/relocatable_shaders/DescBufferReloc.vert).
+
+Note that for the top level case, it is assumed that the descriptor has been
+spilt. The code is generated this way in
+[`DescBuilder::CreateLoadBufferDesc`](../lgc/builder/DescBuilder.cpp). The case
+where `node == topeNode` cannot be reached for unlinked shaders with missing
+user data nodes. Then when replacing the `reloc::DescriptorUseSpillTable`
+relocation in [`RelocHandler::getValue`](../lgc/elfLinker/RelocHandler.cpp), the
+descriptor is spilled if it is a top level node.
+
+Also note that it reserves two input registers, one of which is linked to
+descriptor set 0. That register entry does not contain an offset into user data
+nodes as PAL expected. During the link step, the value will be replaced with the
+offset of the descriptor table that contains descriptor set 0. This happens in
+[`PalMetadata::fixUpRegisters`](../lgc/state/PalMetadata.cpp).
+
+## Compact buffer descriptors
+
+While building an unlinked shader, it is not possible to know if the compact
+buffer descriptors will be used or not. It is unclear if it would be worth
+generating code general enough to work in both cases. The extra overhead would
+potentially negate any gains from using compact buffer descriptors. Therefore
+unlinked shaders are compiled assuming that there are no compact buffer
+descriptors. If the XGL option `EnableRelocatableShaders` is true, then the
+driver will not use compact buffer descriptors.
+
+## Examples
+
+The
+[relocatable shader test directory](../llpc/test/shaderdb/relocatable_shaders)
+contains sample pipelines that use relocatable shaders.
+
+# Performance considerations
+
+Naturally, using relocatable shaders will reduce the quality of the code that is
+generated because the code has to be more general. It is possible that this will
+affect performance, but this will be debilitating. This should be taken into
+consideration when deciding to use relocatable shaders.
+
+Here is a list of known reasons for potential slowdowns in relocatable shaders:
+
+-   Vertex shader prologues: All vertex attributes will have to be in registers
+    at the same time for them to be passed to the body of the shader. This could
+    increase register usage. This to be a bigger problem for smaller shader that
+    use fewer registers.
+
+-   Unused outputs: When compiling a whole pipeline the compiler can look at the
+    fragment shader to identify vertex shader outputs that are unused. Those
+    exports and the code computing the them can be removed. This is not done
+    with unlinked shader because the fragment shader is unavailable when
+    compiling the vertex shader.
```

Comment:
Also packing of vertex shader outputs can't be done.

Reply:
What is meant by packing?  I don't think I have looking into it yet, but we could merge two vec2 outputs into a single vec4 output as long as the decision is made based only on the data types coming from the spir-v.  I don't remember turning anything like that off.